File: builder/code.cpp
#include <iostream>
#include <string>
using namespace std;

// Product class
class Burger {
    string size;
    bool cheese;
    bool pepperoni;
    bool lettuce;

public:
    Burger(string size, bool cheese, bool pepperoni, bool lettuce)
        : size(size), cheese(cheese), pepperoni(pepperoni), lettuce(lettuce) {}

    void describe() {
        cout << "Burger Size: " << size
             << ", Cheese: " << (cheese ? "Yes" : "No")
             << ", Pepperoni: " << (pepperoni ? "Yes" : "No")
             << ", Lettuce: " << (lettuce ? "Yes" : "No") << endl;
    }
};

// Builder class
class BurgerBuilder {
    string size;
    bool cheese = false;
    bool pepperoni = false;
    bool lettuce = false;

public:
    BurgerBuilder(string size) : size(size) {}

    BurgerBuilder& addCheese() {
        cheese = true;
        return *this;
    }

    BurgerBuilder& addPepperoni() {
        pepperoni = true;
        return *this;
    }

    BurgerBuilder& addLettuce() {
        lettuce = true;
        return *this;
    }

    Burger build() {
        return Burger(size, cheese, pepperoni, lettuce);
    }
};

// Client code
int main() {
    Burger burger = BurgerBuilder("Large")
                        .addCheese()
                        .addPepperoni()
                        .build();
    burger.describe();
    return 0;
}

File: builder/code.ts
// Product class
class Burger {
  private size: string;
  private cheese: boolean;
  private tomato: boolean;
  private lettuce: boolean;

  constructor(builder: BurgerBuilder) {
    this.size = builder.size;
    this.cheese = builder.cheese;
    this.tomato = builder.tomato;
    this.lettuce = builder.lettuce;
  }

  describe() {
    console.log(
      `Burger Size: ${this.size}, Cheese: ${this.cheese}, Tomato: ${this.tomato}, Lettuce: ${this.lettuce}`
    );
  }
}

// Builder class
class BurgerBuilder {
  size: string;
  cheese: boolean = false;
  tomato: boolean = false;
  lettuce: boolean = false;

  constructor(size: string) {
    this.size = size;
  }

  addCheese(): BurgerBuilder {
    this.cheese = true;
    return this;
  }

  addTomato(): BurgerBuilder {
    this.tomato = true;
    return this;
  }

  addLettuce(): BurgerBuilder {
    this.lettuce = true;
    return this;
  }

  build(): Burger {
    return new Burger(this);
  }
}

// Client code
const burger = new BurgerBuilder("Large")
  .addCheese()
  .addTomato()
  .build();

burger.describe();
// Output: Burger Size: Large, Cheese: true, Pepperoni: true, Lettuce: false
